consensus/src/dag/tests/integration_tests.rs

```diff
@@ -0,0 +1,232 @@
+) -> (
+    Vec<DagBootstrapUnit>,
+    Vec<UnboundedReceiver<Vec<Arc<CertifiedNode>>>>,
+) {
+    let peers_and_metadata = playground.peer_protocols();
+    let (nodes, ordered_node_receivers) = signers
+        .iter()
+        .enumerate()
+        .map(|(id, signer)| {
+            let peer_id = signer.author();
+            let mut conn_meta = ConnectionMetadata::mock(peer_id);
+            conn_meta.application_protocols = ProtocolIdSet::from_iter([
+                ProtocolId::ConsensusDirectSendJson,
+                ProtocolId::ConsensusDirectSendBcs,
+                ProtocolId::ConsensusRpcBcs,
+            ]);
+            let peer_network_id = PeerNetworkId::new(NetworkId::Validator, peer_id);
+            peers_and_metadata
+                .insert_connection_metadata(peer_network_id, conn_meta)
+                .unwrap();
+
+            let (_, storage) = MockStorage::start_for_testing((&validators).into());
+            let (network, network_events) =
+                create_network(playground, id, signer.author(), validators.clone());
+
+            DagBootstrapUnit::new(
+                signer.author(),
+                1,
+                signer.clone(),
+                storage,
+                network,
+                Arc::new(ClockTimeService::new(runtime.handle().clone())),
+                aptos_time_service::TimeService::real(),
+                network_events,
+            )
+        })
+        .unzip();
+
+    (nodes, ordered_node_receivers)
+}
+
+#[tokio::test]
+async fn test_dag_e2e() {
+    let num_nodes = 7;
+    let runtime = consensus_runtime();
+    let mut playground = NetworkPlayground::new(runtime.handle().clone());
+    let (signers, validators) = random_validator_verifier(num_nodes, None, false);
+    let author_indexes = validators.address_to_validator_index().clone();
+
+    let (nodes, mut ordered_node_receivers) =
+        bootstrap_nodes(&mut playground, &runtime, signers, validators);
+    for node in nodes {
+        runtime.spawn(node.start());
+    }
+
+    runtime.spawn(playground.start());
+
+    let display = |node: &Arc<CertifiedNode>| {
+        (
+            node.metadata().round(),
+            *author_indexes.get(node.metadata().author()).unwrap(),
+        )
+    };
+
+    for _ in 1..10 {
+        let mut all_ordered = vec![];
+        for receiver in &mut ordered_node_receivers {
+            let block = receiver.next().await.unwrap();
+            all_ordered.push(block)
+        }
+        let first: Vec<_> = all_ordered
+            .iter()
+            .next()
+            .unwrap()
+            .iter()
+            .map(display)
+            .collect();
+        assert_gt!(first.len(), 0, "must order nodes");
+        debug!("Nodes: {:?}", first);
+        for ordered in all_ordered.iter() {
+            let a: Vec<_> = ordered.iter().map(display).collect();
+            assert_gt!(a.len(), 0, "must order nodes");
+            assert_eq!(a, first[..a.len()]);
```

Comment:
I don't see it's guaranteed `first` has more elements than `a`?

Reply:
I was assuming they all will be of the same length, because we don't have any faults. Changed the assertion to check the length matches and the contents to be precise.